### DISCUSSION TITLE
Significant delay when directing to getStaticProps() page (it waits for data fetching)

### DISCUSSION THREAD

**Discussion by AshConnolly** [vercel/next.js#14685] [Help]
Hi there! 😃 👋

When clicking a link that redirects to a page that uses `getStaticProps()` there is a significant delay before the user see's any changes. This is because it's waiting for all the data fetching, in the target page's `getStaticProps()` to finish downloading, before redirecting the user.

**Is this expected behaviour?** 

How would one go about redirecting immediately and showing the loading state of the target page? A simple anchor tag and href works but there are big performance downsides to it. I can't find anything in the docs about this.

In addition to this, the page does not scroll to the top, neither using Link/router or with an anchor tag. As I understand it the default behaviour should scroll to top.

Behaviour:
- user visits a page using incremental static generation 
- users sees 'Loading...'
- page content loads and is shown
- user clicks any of the links to redirect to another page on the same route (incremental static generation)
- user has to wait, with no UI feedback that they are being redirected
- the target page `getStaticProps` finishes downloading content
- user is redirected and UI is updated

This happens if coming to this page from any location.

Example code below. Artificial delay has been added:
```js
// github/[user].js
import { useRouter } from 'next/router'
import Link from 'next/link'

function UserPage(props) {
  const router = useRouter()
  var url = 'github'

  if (router.isFallback) {
    return <div>Loading...</div>
  }
  const { name, bio, avatar_url } = props.userData
  // prettier-ignore
  const style = () => <style jsx>{`a {display: block;}`}</style>

  return (
    <div>
      <img src={avatar_url} alt="" />
      <h2>{name}</h2>
      <p>{bio}</p>

      <Link href={`/github/[user]`} as={`/github/rauchg`}>
        <a>rauchg</a>
      </Link>
      <Link href={`/github/[user]`} as={`/github/timneutkens`}>
        <a>timneutkens</a>
      </Link>
      <Link href={`/${url}/[user]`} as={`/${url}/timer`}>
        <a>Joe Haddad</a>
      </Link>

      <a onClick={() => router.push(`/${url}/[user]`, `/${url}/timneutkens`)}>timneutkens</a>
      <a onClick={() => router.push(`/${url}/[user]`, `/${url}/rauchg`)}>rauchg</a>
      <a onClick={() => router.push(`/${url}/[user]`, `/${url}/timer`)}>Joe Haddad</a>

      {style()}
    </div>
  )
}

export async function getStaticPaths() {
  return {
    paths: [],
    fallback: true,
  }
}

export async function getStaticProps({ params }) {
  const delay = ms => new Promise(resolve => setTimeout(resolve, ms))
  await delay(3000).then(() => console.log('waited 3 seconds'))
  let userData = await fetch('https://api.github.com/users/' + params.user).then(r => r.json())
  return { props: { userData } }
}

export default UserPage
```

**Comment by AshConnolly** [comment]
Anyone got any ideas? Still unaware how to solve this problem, and the documentation has no guidance.

Any help would be appreciated! 😃

**Comment by jamesmosier** [comment]
You could use something similar to the [`with-loading`](https://github.com/vercel/next.js/blob/canary/examples/with-loading/pages/_app.js#L6-L11) example:

```js
useEffect(() => {
  const handleRouteChange = (url) => {
    console.log('App is changing to: ', url)
    // show spinner if not `router.isFallback`
  }

  const handleRouteChangeComplete = () => {
    // hide spinner
  }

  Router.events.on('routeChangeStart', handleRouteChange)
  Router.events.on('routeChangeComplete', handleRouteChangeComplete)
  return () => {
    Router.events.off('routeChangeStart', handleRouteChange)
    Router.events.off('routeChangeComplete', handleRouteChangeComplete)
  }
}, [])
```

**Reply by AshConnolly** [reply]
Thanks for the help James! 😃
That does look interesting, but in this case I'm just wanting to send users to the destination page right away, and have them see the destination page in a loading state via router.isFallback. 

As opposed to them clicking a link on the source page, and then having to wait for all destination page data fetching to complete, then sending them to it. 

I'd describe this as normal router behaviour, seems odd to me how it's not the default 😕

**Comment by sanderkranz** [comment]
Found in [docs](https://nextjs.org/docs/basic-features/data-fetching/get-static-props#runs-on-every-request-in-development):
> In development (next dev), getStaticProps will be called on every request.

Does it make sense?

**Reply by illiteratewriter** [reply]
I think I am having the issue in production also.

**Comment by veewoo** [comment]
Up

**Comment by cacoos** [comment]
Happening to me also!

**Comment by devlaza95** [comment]
This also happens to me. I have a blog page wired up with Sanity CMS. I only have one blog article in the dataset with some dummy text.

The redirect (response) took 133-140ms. 😕

Anybody managed to figure this out?
